src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs

```diff
@@ -163,43 +163,70 @@ SymbolKind.NetModule or
+                    // 2. The document wasn't present in the original snapshot at all, and we just need to add the tree.
+                    // 3. The document wasn't present in the original snapshot, but an older file had been removed that had the same file path.
+                    //    As a heuristic, we remove the old one so we don't end up with duplicate trees.
+                    //
+                    // Note it's possible that we simply had never tried to produce a compilation yet for this project at all, in that case
+                    // GetPartialCompilationState would have produced an empty compilation, and it would have updated inProgressProject to
+                    // remove all the documents. Thus, that is no different than the "add" case above.
+
+                    if (inProgressProject.DocumentStates.TryGetState(docState.Id, out var oldState))
                     {
-                        compilationPair = compilationPair.ReplaceSyntaxTree(existingTree, tree);
+                        // Scenario 1. The document had been previously parsed and it's there, so we can update it with our current state
+                        // This call should be instant, since the compilation already must exist that contains this tree. Note if no compilation existed
+                        // GetPartialCompilationState would have produced an empty one, and removed any documents, so inProgressProject.DocumentStates would
+                        // have been empty originally.
+                        var oldTree = oldState.GetSyntaxTree(cancellationToken);
+
+                        compilationPair = compilationPair.ReplaceSyntaxTree(oldTree, tree);
                         inProgressProject = inProgressProject.UpdateDocument(docState, textChanged: false, recalculateDependentVersions: false);
                     }
                     else
                     {
-                        compilationPair = compilationPair.AddSyntaxTree(tree);
-                        Debug.Assert(!inProgressProject.DocumentStates.Contains(docState.Id));
-                        inProgressProject = inProgressProject.AddDocuments(ImmutableArray.Create(docState));
+                        // We're in either scenario 2 or 3. Do we have an existing tree to try replacing?
+                        var oldTree = compilationPair.CompilationWithoutGeneratedDocuments.SyntaxTrees.FirstOrDefault(t => t.FilePath == tree.FilePath);
+                        if (oldTree == null)
+                        {
+                            // Scenario 2.
+                            compilationPair = compilationPair.AddSyntaxTree(tree);
+                            inProgressProject = inProgressProject.AddDocuments(ImmutableArray.Create(docState));
+                        }
+                        else
+                        {
+                            // Scenario 3.
```

Comment:
Consider adding comment to explain the nuances of documents with null path and how it falls into either scenario 2 or 3.